client/gc_test.go

```diff
@@ -290,40 +290,110 @@ func TestAllocGarbageCollector_MakeRoomForAllocations_GC_Fallback(t *testing.T)
 	}
 }
 
+// TestAllocGarbageCollector_MakeRoomForAllocations_MaxAllocs asserts that when
+// making room for new allocs, terminal allocs are GC'd until old_allocs +
+// new_allocs <= limit
 func TestAllocGarbageCollector_MakeRoomForAllocations_MaxAllocs(t *testing.T) {
 	t.Parallel()
-	const (
-		liveAllocs   = 3
-		maxAllocs    = 6
-		gcAllocs     = 4
-		gcAllocsLeft = 1
-	)
-
-	logger := testLogger()
-	statsCollector := &MockStatsCollector{
-		availableValues: []uint64{10 * 1024 * MB},
-		usedPercents:    []float64{0},
-		inodePercents:   []float64{0},
-	}
-	allocCounter := &MockAllocCounter{allocs: liveAllocs}
-	conf := gcConfig()
-	conf.MaxAllocs = maxAllocs
-	gc := NewAllocGarbageCollector(logger, statsCollector, allocCounter, conf)
-
-	for i := 0; i < gcAllocs; i++ {
-		_, ar := testAllocRunnerFromAlloc(mock.Alloc(), false)
-		close(ar.waitCh)
-		gc.MarkForCollection(ar)
-	}
-
-	if err := gc.MakeRoomFor([]*structs.Allocation{mock.Alloc(), mock.Alloc()}); err != nil {
-		t.Fatalf("error making room for 2 new allocs: %v", err)
-	}
-
-	// There should be gcAllocsLeft alloc runners left to be collected
-	if n := len(gc.allocRunners.index); n != gcAllocsLeft {
-		t.Fatalf("expected %d remaining GC-able alloc runners but found %d", gcAllocsLeft, n)
-	}
+	assert := assert.New(t)
+	server, serverAddr := testServer(t, nil)
+	defer server.Shutdown()
+	testutil.WaitForLeader(t, server.RPC)
+
+	const maxAllocs = 6
+	client := testClient(t, func(c *config.Config) {
+		c.GCMaxAllocs = maxAllocs
+		c.RPCHandler = server
+		c.Servers = []string{serverAddr}
+		c.ConsulConfig.ClientAutoJoin = new(bool) // squelch logs
+	})
+	defer client.Shutdown()
+	waitTilNodeReady(client, t)
+
+	assertAllocs := func(expectedAll, expectedDestroyed int) {
+		// Wait for allocs to be started
+		testutil.WaitForResult(func() (bool, error) {
+			all, destroyed := 0, 0
+			for _, ar := range client.getAllocRunners() {
+				all++
+				if ar.IsDestroyed() {
+					destroyed++
+				}
+			}
+			return all == expectedAll && destroyed == expectedDestroyed, fmt.Errorf(
+				"expected %d allocs (found %d); expected %d destroy (found %d)",
+				expectedAll, all, expectedDestroyed, destroyed,
+			)
+		}, func(err error) {
+			t.Fatalf("alloc state: %v", err)
+		})
+	}
+
+	// Create a job
+	state := server.State()
+	job := mock.Job()
+	job.TaskGroups[0].Tasks[0].Driver = "mock_driver"
+	job.TaskGroups[0].Tasks[0].Config["run_for"] = "30s"
+	nodeID := client.Node().ID
+	if err := state.UpsertJob(98, job); err != nil {
```

Comment:
Use the assert library

Reply:
I did at first and then got really confusing test failures. I didn't realize The assert library uses `t.Error` instead of `t.Fatal` and you need to check its return bool to know whether or not to fail!

So I just skipped the assert library as it didn't seem to add anything.